### DISCUSSION TITLE
Allow Tasker to read habit data

### DISCUSSION THREAD

**Discussion by Erudition** [iSoron/uhabits#423] [Feature Requests]
While I suppose it would be nice to have a habit automatically checked off when something happens (as is currently possible with the check habit action in Tasker), it's not really relevant in the critical process of _forming_ habits. Where Tasker automation would really be powerful is in that area: "if I haven't done _xyz_ yet today, nag me (with some set of actions) until I have". This could make the difference between something I occasionally check off when I remember, and a truly lasting habit. The built-in reminders alone are not effective enough for me.

**Unfortunately, the only Tasker "integration" currently is one-way (write-only), allowing automated "checking" of habits, but no ability to see whether one is checked or not.**

Without this feature, the best one can do (that I can imagine) is enable early reminder notifications for all relevant habits, make them persistent so they can only go away by being checked, and then have Tasker read the system notifications to see which habits are not done. This might work, but would leave a mess of permanent notifications all day.

It seems the lead developer is concerned about implementing this securely. That's also a discussion to be had, although in the interim, I personally don't care if all my device's other apps can see that I flossed this morning.

**Comment by iSoron** [comment]
I agree that this is an important feature, and I would be very happy to review pull requests implementing this feature. 

From the feedback I received, however, I know for a fact that many users are using the app to track very sensitive habits (for example, related to medical conditions or addictions). For this reason, I would only accept such a pull request if security measures were put in place to prevent other malicious apps from extracting that information.

**Comment by Erudition** [comment]
Okay, then how about an app setting that's simply "Allow other apps to read my habit data", disabled by default?

That is, if it's not simply possible to just restrict access to exclusively automation apps like Tasker. Which would be the more obvious approach, so I'm assuming you can't do that?

**Comment by digitalcircuit** [comment]
In the meantime, if you have root access or (untested) ADB WiFi set up, it's possible to hack around this in Tasker by searching for the notifications posted.

```
dumpsys notification | grep -e "NotificationRecord(0x.*: pkg=org.isoron.uhabits user=UserHandle{[0-9]*} id=[0-9]*.*Notification(channel=REMINDERS" || echo absent
```

This will either print the notifications from Loop, or `absent` if none are currently posted.  I haven't extended this to determine *which* habits are currently in the notification list.

Similar, Tasker can use notification permissions to listen for the `Notification Removed` event to determine when an event is checked or dismissed.

This is far from ideal, but it's the closest I've gotten without trying using root to directly read Loop's habit database.

*Motivation: I've implemented a system in Tasker that will automatically enable Digital Wellbeing's Focus Mode once Sleep as Android starts an alarm, blocking social media and web browsing apps, then once I finish my morning habits by checking them off (clearing the notification), Focus Mode will be automatically disabled.*

I'd be in favor of `Allow other apps to read my habit data`, similar to how [Sleep as Android lets you turn Tasker integration on/off](https://sleep.urbandroid.org/ufaqs/tasker-integration-not-work/ ) (alongside [another tutorial](https://sleep.urbandroid.org/documentation/tutorials/tasker/ )).

If possible, it'd be nice to allow other automation apps beyond Tasker to read the data, e.g. [Easer, an open-source Android automation app](https://me.ryey.icu/Easer/en/ ).

**Reply by Erudition** [reply]
That's actually what I did as well!
Funny enough, I also have my Sleep As Android rigged to start a DW Focus Mode (albeit on dismiss, not start) via AutoInput (since only Bixby can automate DW apparently) though I do have root so I'm curious how you did it.

**Reply by digitalcircuit** [reply]
@Erudition If you're asking about automating Digital Wellbeing, it's a bit of a kludge but it works:
https://android.stackexchange.com/questions/214419/how-can-i-emulate-pressing-another-applications-quick-settings-tile/221910#221910

Over Android updates I've had to change whether or not the screen is on, and whether or not I actually expand the Quick Settings page.  Currently, on February LineageOS 20 builds, I can toggle without expanding the QS tiles, but I think the screen might need to be on.

(My Tasker script waits for the device to be unlocked before toggling just so I don't have to worry about this in the future.)

**Reply by Erudition** [reply]
Oh yeah, I literally had it opening DW and tapping into Focus Mode menus, after turning on and unlocking the screen, quick tile sounds way simpler. I do like that I was able to natively give it a 2-hr limit though. Nice!

**Comment by iSoron** [comment]
> Okay, then how about an app setting that's simply "Allow other apps to read my habit data", disabled by default?

I think that would be fine. If anyone is interested in implementing this, feel free to submit a pull request.

**Comment by mshuflin** [comment]
I would also like this feature

**Comment by tjthejuggler** [comment]
Does anyone know if this ended up happening?

**Reply by Erudition** [reply]
Not until someone implements it and submits a PR, which no one has volunteered to do so far.

**Comment by Spiralo-Idioide** [comment]
If rooted it is also possible to do an SQL query using tasker SQL Query action or termux sqlite3 cmd. The relevant file should be `/data/user/0/org.isoron.uhabits/databases/uhabits.db`
